File: server/internal/services/faction_service.go
package services

import (
	"errors"
	"warhammer-api/internal/models"
	"warhammer-api/internal/repositories"
)

type FactionService struct {
	repo repositories.FactionRepository
}

func NewFactionService(repo repositories.FactionRepository) *FactionService {
	return &FactionService{repo}
}

func (s *FactionService) GetAll() ([]models.Faction, error) {
	return s.repo.GetAll()
}

func (s *FactionService) GetByID(id int) (*models.Faction, error) {
	return s.repo.GetById(id)
}

func (s *FactionService) Create(data *models.Faction) (*models.Faction, error) {
	if data.Name == "" {
		return nil, errors.New("name cannot be empty")
	}
	return s.repo.Create(data)
}

func (s *FactionService) Update(data *models.Faction) (*models.Faction, error) {
	return s.repo.Update(data)
}

func (s *FactionService) Delete(id int) error {
	return s.repo.Delete(id)
}

File: server/services/factionService.js
const factionRepository = require("../repositories/factionRepository");
const { createLegacyNotFoundError } = require("../lib/legacyApi");

class FactionService {
  async getAll(options = {}) {
    return await factionRepository.findAll(options);
  }

  async getById(id) {
    const faction = await factionRepository.findById(id);
    if (!faction) {
      throw createLegacyNotFoundError("Faction");
    }
    return faction;
  }

  async create(factionData) {
    return await factionRepository.create(factionData);
  }

  async update(id, factionData) {
    const updated = await factionRepository.update(id, factionData);
    if (!updated) {
      throw createLegacyNotFoundError("Faction");
    }
    return updated;
  }

  async delete(id) {
    const deleted = await factionRepository.delete(id);
    if (!deleted) {
      throw createLegacyNotFoundError("Faction");
    }
    return deleted;
  }
}

module.exports = new FactionService();
